pkg/resourcemanager/psql/psqluser/psqluser.go

```diff
@@ -0,0 +1,285 @@
+	psdatabase "github.com/Azure/azure-service-operator/pkg/resourcemanager/psql/database"
+	"github.com/Azure/azure-service-operator/pkg/secrets"
+
+	"github.com/Azure/azure-service-operator/api/v1alpha1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	_ "github.com/lib/pq" //the pg lib
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// PSqlServerPort is the default server port for sql server
+const PSqlServerPort = 5432
+
+// PDriverName is driver name for psqldb connection
+const PDriverName = "postgres"
+
+// PSecretUsernameKey is the username key in secret
+const PSecretUsernameKey = "username"
+
+// PSecretPasswordKey is the password key in secret
+const PSecretPasswordKey = "password"
+
+//PostgreSqlUserManager for psqluser manager
+type PostgreSqlUserManager struct {
+	SecretClient secrets.SecretClient
+	Scheme       *runtime.Scheme
+}
+
+//NewPostgreSqlUserManager creates a new PostgreSqlUserManager
+func NewPostgreSqlUserManager(secretClient secrets.SecretClient, scheme *runtime.Scheme) *PostgreSqlUserManager {
+	return &PostgreSqlUserManager{
+		SecretClient: secretClient,
+		Scheme:       scheme,
+	}
+}
+
+// GetDB retrieves a database
+func (s *PostgreSqlUserManager) GetDB(ctx context.Context, resourceGroupName string, serverName string, databaseName string) (db psql.Database, err error) {
+	dbClient, err := psdatabase.GetPSQLDatabasesClient()
+	if err != nil {
+		return psql.Database{}, err
+	}
+
+	return dbClient.Get(
+		ctx,
+		resourceGroupName,
+		serverName,
+		databaseName,
+	)
+}
+
+// ConnectToSqlDb connects to the PostgreSQL db using the given credentials
+func (s *PostgreSqlUserManager) ConnectToSqlDb(ctx context.Context, drivername string, server string, database string, port int, user string, password string) (*sql.DB, error) {
+
+	psqldbdnssuffix := "database.azure.com"
+	if config.Environment().Name != "AzurePublicCloud" {
+		psqldbdnssuffix = config.Environment().SQLDatabaseDNSSuffix
+	}
+	//the host or fullserveraddress should be:
+	//for public cloud <server>.postgres.database.azure.com
+	//for China cloud <server>.postgres.database.chinacloudapi.cn
+	//for German cloud <server>.postgres.database.cloudapi.de
+	//for US government <server>.postgres.database.usgovcloudapi.net
+	fullServerAddress := fmt.Sprintf("%s.postgres."+psqldbdnssuffix, server)
+
+	connString := fmt.Sprintf("host=%s user=%s password=%s port=%d dbname=%s sslmode=require connect_timeout=30", fullServerAddress, user, password, port, database)
+
+	db, err := sql.Open(drivername, connString)
+	if err != nil {
+		return db, err
+	}
+
+	err = db.PingContext(ctx)
+	if err != nil {
+		return db, err
+	}
+
+	return db, err
+}
+
+// GrantUserRoles grants roles to a user for a given database
+func (s *PostgreSqlUserManager) GrantUserRoles(ctx context.Context, user string, roles []string, db *sql.DB) error {
+	var errorStrings []string
+	for _, role := range roles {
+		tsql := fmt.Sprintf("GRANT %s TO %q", role, user)
```

Comment:
You should use the interpolation built into the db client to dynamically craft sql queries. This is the only way to avoid sql injection. Some places in Azure SQL Server it is impossible so we have a workaround for one or two queries. In this case I believe the normal path should work fine.

e.g. https://github.com/Azure/azure-service-operator/blob/master/pkg/resourcemanager/azuresql/azuresqluser/azuresqluser.go#L154

Reply:
Looks like some PSQL commands like CREATE USER and possible GRANT ROLE can't use the normal string interpolation to avoid sql injection. That means you will want to use the same filtering used in the CREATE USER command on the variables here.